### PR TITLE
Custom preauthorize annotation

### PR DESCRIPTION
**Pull request by OH72**
dev
## JIRA

* [Main JIRA ticket](https://jira.softserve.academy/secure/RapidBoard.jspa?rapidView=id)

## Code reviewers

- [ ] @github_username

### Second Level Review

- [ ] @github_username

## Summary of issue

ToDo

## Summary of change

PreAuthorize needs constatnt value or T("path to package").ROLE. First way isnt apropriate because we want to use enum. Second way makes annotation very long and unreadable. Threfore I created own annotation, which works with enum (@AllowedRoles({RoleData.ADMIN, RoleData.MANAGER}) ).

## Testing approach

ToDo

## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=95%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

### CODE REVIEW

File: src/main/java/com/softserve/teachua/controller/AboutUsItemController.java [hunk 1]
```diff
@@ -36,7 +38,7 @@ public AboutUsItemResponse getAboutUsItems(
         return aboutUsItemService.getAboutUsItemResponseById(id);
     }
 
-    @PreAuthorize("hasRole(T(com.softserve.teachua.constants.RoleData).ADMIN.getDBRoleName())")
+    @AllowedRoles(RoleData.ADMIN)
```
Comment (OH72) [inline src/main/java/com/softserve/teachua/controller/AboutUsItemController.java:41]: It is custom annotation, see below

File: src/main/java/com/softserve/teachua/utils/annotation/AllowedRoles.java [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package com.softserve.teachua.utils.annotation;
+
+import com.softserve.teachua.constants.RoleData;
+
+import java.lang.annotation.*;
+
+@Documented
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.METHOD})
+public @interface AllowedRoles {
```
Comment (OH72) [inline src/main/java/com/softserve/teachua/utils/annotation/AllowedRoles.java:10]: annotation interface

File: src/main/java/com/softserve/teachua/utils/annotation/AllowedRolesAspect.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+@Aspect
+@Component
+@Slf4j
+public class AllowedRolesAspect {
+
+//    private static final String ACCESS_DENIED_EXCEPTION = "do not have permission";
+
+    @Around("@annotation(com.softserve.teachua.utils.annotation.AllowedRoles)")
+    public Object doSomething(ProceedingJoinPoint jp) throws Throwable {
+        Set<RoleData> roles = Arrays.stream(((MethodSignature) jp.getSignature()).getMethod()
```
Comment (OH72) [inline src/main/java/com/softserve/teachua/utils/annotation/AllowedRolesAspect.java:28]: get set of allowed roles

File: src/main/java/com/softserve/teachua/utils/annotation/AllowedRolesAspect.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+                .getAnnotation(AllowedRoles.class).value()).collect(Collectors.toSet());
+        log.debug("Allowed roles: {}", roles);
+        HttpServletRequest httpServletRequest = getRequest();
+        for(RoleData role : roles){
+            if(httpServletRequest.isUserInRole(role.getDBRoleName())){
```
Comment (OH72) [inline src/main/java/com/softserve/teachua/utils/annotation/AllowedRolesAspect.java:33]: check if at least one permitted role is equal to user role

File: src/main/java/com/softserve/teachua/utils/annotation/AllowedRolesAspect.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+                return jp.proceed();
+            }
+        }
+
+        throw new AccessDeniedException("");
```
Comment (OH72) [inline src/main/java/com/softserve/teachua/utils/annotation/AllowedRolesAspect.java:38]: throw Forbidden 403 (@PreAuthorize also throws this exception)

File: src/main/java/com/softserve/teachua/utils/annotation/AllowedRolesAspect.java [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+//        throw new WrongAuthenticationException(ACCESS_DENIED_EXCEPTION);
+    }
+
+    private HttpServletRequest getRequest() {
```
Comment (OH72) [inline src/main/java/com/softserve/teachua/utils/annotation/AllowedRolesAspect.java:42]: retunr HttpServletRequest of current user

**Review by VictoriaAmazonka — APPROVED** [review]

**Pull request merged**